src/rgw/rgw_op.cc

```diff
@@ -6397,10 +6397,40 @@ void RGWCompleteMultipart::execute(optional_yield y)
     return;
   }
 
-  op_ret = upload->complete(this, y, s->cct, parts->parts, remove_objs, accounted_size, compressed, cs_info, ofs, s->req_id, s->owner, olh_epoch, s->object.get());
-  if (op_ret < 0) {
-    ldpp_dout(this, 0) << "ERROR: upload complete failed ret=" << op_ret << dendl;
-    return;
+  while (true) {
+    RGWObjVersionTracker objv_tracker = meta_obj->get_version_tracker();
+
+    op_ret = upload->complete(this, y, s->cct, parts->parts, remove_objs, accounted_size, compressed, cs_info, ofs, s->req_id, s->owner, olh_epoch, s->object.get());
+    if (op_ret < 0) {
+      ldpp_dout(this, 0) << "ERROR: upload complete failed ret=" << op_ret << dendl;
+      return;
+    }
+
+    // remove the upload meta object ; the meta object is not versioned
+    // when the bucket is, as that would add an unneeded delete marker
+    int ret = meta_obj->delete_object(this, y, rgw::sal::FLAG_PREVENT_VERSIONING, &objv_tracker);
+    if (ret != -ECANCELED) {
+      if (ret >= 0) {
+        /* serializer's exclusive lock is released */
+        serializer->clear_locked();
+      } else {
+        ldpp_dout(this, 1) << "ERROR: failed to remove object " << meta_obj << ", ret: " << ret << dendl;
+      }
+      break;
+    }
+
+    ldpp_dout(this, 20) << "deleting meta_obj is cancelled due to mismatch cls_version: " << objv_tracker << dendl;
+    meta_obj->get_version_tracker().clear();
+    ret = meta_obj->get_obj_attrs(s->yield, this);
```

Comment:
> this is using `get_obj_attrs()` to read the updated cls version? i was thinking we'd do that in `RadosMultipartUpload::list_parts()`, but maybe this is good enough

I believed `RadosMultipartUpload::list_parts()` was a better place too at the beginning, but realized that I'll have to go along the system obj line in order to find the [right ObjectReadOperation](https://github.com/ceph/ceph/blob/main/src/rgw/services/svc_sys_obj_core.cc#L312) to set `cls_versrion_read` on. It's not something wrong, but not sure if we want to change RGWSI_SysObj interface for this sole purpose, since we already have the logic to bring the `cls_version` back along the `rgw_sal_rados` and `rgw_rados` line? Any other use cases to get the `cls_version` for system object?
In this particular case, adding one more osd trip to get `cls_version` back in racing condition doesn't seem to be a big deal for a multipart complete request.